pkg/processor/eventprocessor/pool.go

```diff
@@ -175,71 +162,76 @@ func (sa *syncPoolAllocator) GetStatistics() *statistics.AllocatorStatistics {
 	return allocatorStatistics
 }
 
-func (sa *syncPoolAllocator) SignalDraining() error {
-	errGroup, _ := errgroup.WithContext(context.Background(), sa.logger)
+type nonBlockingPoolAllocator struct {
+	*abstractPoolAllocator
 
-	for _, objectInstance := range sa.GetObjects() {
-		objectInstance := objectInstance
+	// index used for round-robin allocation
+	index atomic.Uint64
+}
 
-		errGroup.Go(fmt.Sprintf("Drain object %d", objectInstance.GetIndex()), func() error {
-			// if object is not already drained, signal it to drain events
-			if err := objectInstance.Drain(); err != nil {
-				return errors.Wrapf(err, "Failed to signal object %d to drain events", objectInstance.GetIndex())
-			}
-			return nil
-		})
+func NewNonBlockingPoolAllocator(parentLogger logger.Logger, processors []EventProcessor) (Allocator, error) {
+	nonBlockingPoolAllocatorInstance := &nonBlockingPoolAllocator{
+		abstractPoolAllocator: newAbstractPoolAllocator(parentLogger.GetChild("nonblock-pool-allocator")),
+		index:                 atomic.Uint64{},
 	}
-
-	if err := errGroup.Wait(); err != nil {
-		return errors.Wrap(err, "At least one object failed to drain")
+	if err := nonBlockingPoolAllocatorInstance.SetObjects(processors); err != nil {
+		return nil, errors.Wrap(err, "Failed to set non blocking pool allocator")
 	}
 
-	return nil
+	return nonBlockingPoolAllocatorInstance, nil
 }
 
-func (sa *syncPoolAllocator) SignalContinue() error {
-	errGroup, _ := errgroup.WithContext(context.Background(), sa.logger)
+// Allocate allocates an EventProcessor in a non-blocking manner
+func (nba *nonBlockingPoolAllocator) Allocate(timeout time.Duration) (EventProcessor, error) {
+	// Atomically increment and get the index
+	// If idx exceeds math.MaxUint64, it will wrap back to 0, and the subsequent modulo will still yield nba valid slot
+	// For optimal performance, this uses a combined atomic add-and-load operation.
+	// As a result, the first allocated object will have index 1 instead of 0, which is only the case if len(object) >=2
+	// If len(object) == 1, the first object will be allocated with index 0
```

Comment:
I get the logic, but we should never get here if `len(objects)==1`, since in this case we will create an `AsyncSingletonAllocator` (as in `worker/factory.go`)

Reply:
That's right, however this implementation is abstract, so we don't enforce here the number to be more than 1. it is managed outside the implementation.